common/src/View/MapDocument.cpp

```diff
@@ -1124,50 +1125,58 @@ namespace TrenchBroom {
             unused(result);
         }
 
-        void MapDocument::moveLayer(Model::LayerNode* layer, int direction) {
-            if (!canMoveLayer(layer, direction)) {
-                return;
-            }
-            const Transaction transaction(this, "Move Layer");
-
-            ensure(layer != m_world->defaultLayer(), "attempted to move default layer");
-
-            std::vector<Model::LayerNode*> sorted = m_world->customLayersUserSorted();
+        bool MapDocument::moveLayerByOne(Model::LayerNode* layer, MoveDirection direction) {
+            const std::vector<Model::LayerNode*> sorted = m_world->customLayersUserSorted();
 
-            auto maybeIndex = kdl::vec_index_of(sorted, layer);
-            ensure(maybeIndex.has_value(), "layer not found");
-            const int indexInt = static_cast<int>(maybeIndex.value());
-
-            Model::LayerNode* neighbour = sorted.at(static_cast<size_t>(indexInt + direction));
+            const auto maybeIndex = kdl::vec_index_of(sorted, layer);
+            if (!maybeIndex.has_value()) {
+                return false;
+            }
 
+            const int newIndex = static_cast<int>(*maybeIndex) + (direction == MoveDirection::Down ? 1 : -1);
+            if (newIndex < 0 || newIndex >= static_cast<int>(sorted.size())) {
+                return false;
+            }
+            
+            Model::LayerNode* neighbour = sorted.at(static_cast<size_t>(newIndex));           
             const int ourSortIndex = layer->sortIndex();
             const int neighbourSortIndex = neighbour->sortIndex();
-    
+
+            // Swap the sort indices of `layer` and `neighbour`
             executeAndStore(ChangeEntityAttributesCommand::setForNodes({ layer },     Model::AttributeNames::LayerSortIndex, std::to_string(neighbourSortIndex)));
             executeAndStore(ChangeEntityAttributesCommand::setForNodes({ neighbour }, Model::AttributeNames::LayerSortIndex, std::to_string(ourSortIndex)));
+            return true;
+        }
+
+        void MapDocument::moveLayer(Model::LayerNode* layer, const int offset) {
+            ensure(layer != m_world->defaultLayer(), "attempted to move default layer");
+
+            const Transaction transaction(this, "Move Layer");
+            
+            const MoveDirection direction = (offset > 0) ? MoveDirection::Down : MoveDirection::Up;
+            for (int i = 0; i < std::abs(offset); ++i) {
+                if (!moveLayerByOne(layer, direction)) {
```

Comment:
If this returns false, then that indicates that something was quite wrong, doesn't it? I wonder if it's better to log an error and roll back the transaction instead leaving some layers moved and some not?

Reply:
false there either means you tried to move the layer past the end of the list (we clamp movement, there's a test that tries to move by an offset of 1000) or the layer wasn't found.

In either case the layer list is in a consistent state so I think it's fine.. we won't hit these in practice anyway